src/Symfony/Component/String/BinaryString.php

```diff
@@ -0,0 +1,475 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+use Symfony\Component\String\Exception\RuntimeException;
+
+/**
+ * Represents a binary-safe string of bytes.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ *
+ * @experimental in 5.0
+ */
+class BinaryString extends AbstractString
+{
+    public function __construct(string $string = '')
+    {
+        $this->string = $string;
+    }
+
+    public static function fromRandom(int $length = 16): self
+    {
+        $string = '';
+
+        do {
+            $string .= str_replace(['/', '+', '='], '', base64_encode(random_bytes($length)));
+        } while (\strlen($string) < $length);
+
+        return new static(substr($string, 0, $length));
+    }
+
+    public function byteCode(int $offset = 0): ?int
+    {
+        $str = $offset ? $this->slice($offset, 1) : $this;
+
+        return '' === $str->string ? null : \ord($str->string);
+    }
+
+    public function append(string ...$suffix): parent
+    {
+        $str = clone $this;
+        $str->string .= 1 >= \count($suffix) ? ($suffix[0] ?? '') : implode('', $suffix);
+
+        return $str;
+    }
+
+    public function camel(): parent
+    {
+        $str = clone $this;
+        $str->string = lcfirst(str_replace(' ', '', ucwords(preg_replace('/[^a-zA-Z0-9\x7f-\xff]++/', ' ', $this->string))));
+
+        return $str;
+    }
+
+    public function chunk(int $length = 1): array
+    {
+        if (1 > $length) {
+            throw new InvalidArgumentException('The chunk length must be greater than zero.');
+        }
+
+        if ('' === $this->string) {
+            return [];
+        }
+
+        $str = clone $this;
+        $chunks = [];
+
+        foreach (str_split($this->string, $length) as $chunk) {
+            $str->string = $chunk;
+            $chunks[] = clone $str;
+        }
+
+        return $chunks;
+    }
+
+    public function endsWith($suffix): bool
+    {
+        if ($suffix instanceof parent) {
+            $suffix = $suffix->string;
+        } elseif (\is_array($suffix) || $suffix instanceof \Traversable) {
+            return parent::endsWith($suffix);
+        } else {
+            $suffix = (string) $suffix;
+        }
+
+        return \strlen($this->string) - \strlen($suffix) === ($this->ignoreCase ? strripos($this->string, $suffix) : strrpos($this->string, $suffix));
+    }
+
+    public function equalsTo($string): bool
```

Comment:
```suggestion
    /**
     * @param AbstractString|string|string[] $string
     */
    public function equalsTo($string): bool
```
Can be useful for autocompletion, static analysis and so on. Other methods could benefit from this doc.

Reply:
Any objects that implements `__toString()` is allowed actually. That's what `string` means already to me. What's the relation with autocompletion?

Reply:
I mean that when working with PhpStorm it seems that it can help you autocomplete the variable to pass in a function given what's available in the scope if the function is properly documented:
![image](https://user-images.githubusercontent.com/1524501/65674239-841adc80-e04c-11e9-98e6-efda2e82d59d.png)